### PR TITLE
JUCX: support IOV DT.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
Support of IOV data type operation for send/recv tagged and stream.

## Why ?
To match many-to-many API in SparkUCX. 

TODO:

- Switch at some point to nbx API

 @abellina FYI

### CODE REVIEW

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java [hunk 1]
```diff
@@ -235,6 +249,12 @@ public UcpRequest recvStreamNonBlocking(long localAddress, long size, long flags
         return recvStreamNonBlockingNative(getNativeId(), localAddress, size, flags, callback);
     }
 
+    public UcpRequest recvStreamNonBlocking(long[] localAddresses, long[] sizes, long flags,
+                                            UcxCallback callback) {
+        return recvStreamIovNonBlockingNative(getNativeId(),
+            localAddresses, sizes, flags, callback);
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java:255]: wrap line

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java [hunk 2]
```diff
@@ -206,6 +206,15 @@ public UcpRequest sendTaggedNonBlocking(ByteBuffer sendBuffer, UcxCallback callb
...
     }
 
+    /**
+     * Iov version of non blocking send operaation
+     */
+    public UcpRequest sendTaggedNonBlocking(long[] localAddresses, long[] sizes,
+                                            long tag, UcxCallback callback) {
+        return sendTaggedIovNonBlockingNative(getNativeId(),
+            localAddresses, sizes, tag, callback);
+    }
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java:216]: wrap line

File: bindings/java/src/main/native/endpoint.cc [hunk 3]
```diff
@@ -204,6 +204,31 @@ Java_org_openucx_jucx_ucp_UcpEndpoint_sendTaggedNonBlockingNative(JNIEnv *env, j
     return process_request(request, callback);
 }
 
+JNIEXPORT jobject JNICALL
+Java_org_openucx_jucx_ucp_UcpEndpoint_sendTaggedIovNonBlockingNative(JNIEnv *env, jclass cls,
+                                                                    jlong ep_ptr, jlongArray addresses,
+                                                                    jlongArray sizes, jlong tag,
+                                                                    jobject callback)
+{
+    jsize iovcnt = env->GetArrayLength(addresses);
```
Comment (yosefe) [inline bindings/java/src/main/native/endpoint.cc:213]: check that addresses and sizes have same length

File: bindings/java/src/main/native/endpoint.cc [hunk 4]
```diff
@@ -216,6 +241,31 @@ Java_org_openucx_jucx_ucp_UcpEndpoint_sendStreamNonBlockingNative(JNIEnv *env, j
...
 }
 
+JNIEXPORT jobject JNICALL
+Java_org_openucx_jucx_ucp_UcpEndpoint_sendStreamIovNonBlockingNative(JNIEnv *env, jclass cls,
+                                                                     jlong ep_ptr, jlongArray addresses,
+                                                                     jlongArray sizes,
+                                                                     jobject callback)
+{
+    jsize iovcnt = env->GetArrayLength(addresses);
+    ucp_dt_iov_t* iovec = (ucp_dt_iov_t*)ucs_malloc(sizeof(ucp_dt_iov_t) * iovcnt, "JUCX iov vector");
```
Comment (yosefe) [inline bindings/java/src/main/native/endpoint.cc:251]: 1. `sizeof(*iovec) * iovcnt`
2. check for allocation failure

File: bindings/java/src/main/native/endpoint.cc [hunk 5]
```diff
@@ -237,6 +287,40 @@ Java_org_openucx_jucx_ucp_UcpEndpoint_recvStreamNonBlockingNative(JNIEnv *env, j
...
+                                                                     jlong ep_ptr,
+                                                                     jlongArray addresses, jlongArray sizes,
+                                                                     jlong flags, jobject callback)
+{
+    size_t rlength;
+
+    jsize iovcnt = env->GetArrayLength(addresses);
+    ucp_dt_iov_t* iovec = (ucp_dt_iov_t*)ucs_malloc(sizeof(ucp_dt_iov_t) * iovcnt, "JUCX iov vector");
+
+    fill_iov_vector(env, addresses, sizes, iovec, iovcnt);
```
Comment (yosefe) [inline bindings/java/src/main/native/endpoint.cc:301]: maybe unite this to one function: get_ucp_iov(addresses, sizes) -> iovec

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java (hunk 1 above)
Comment (petro-rudenko) [reply bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java:255]: We use 4 spaces in java: https://github.com/openucx/ucx/blob/master/bindings/java/checkstyle.xml#L62 for long methods calls.

Comment (yosefe) [reply bindings/java/src/main/java/org/openucx/jucx/ucp/UcpEndpoint.java:255]: yes, but seems "localAddress" could be on prev line

File: bindings/java/src/main/native/endpoint.cc (hunk 3 above)
Comment (petro-rudenko) [reply bindings/java/src/main/native/endpoint.cc]: Done in java

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java [hunk 6]
```diff
@@ -150,6 +150,12 @@ public UcpRequest recvTaggedNonBlocking(ByteBuffer recvBuffer, UcxCallback callb
     public UcpRequest recvTaggedNonBlocking(long[] localAddresses, long[] sizes,
                                             long tag, long tagMask,
                                             UcxCallback callback) {
+
+        if (localAddresses.length != sizes.length) {
+            throw new UcxException("Arrays of not equal sizes: " +
+                localAddresses.length + " != " + sizes.length);
+        }
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java:157]: use checkArraySizes?

File: bindings/java/src/main/native/jucx_common_def.h [hunk 7]
```diff
@@ -111,10 +111,19 @@ jobject new_tag_msg_instance(JNIEnv *env, ucp_tag_message_h msg_tag,
...
-UCS_F_ALWAYS_INLINE void fill_iov_vector(JNIEnv *env,
-                                         jlongArray addr_array, jlongArray size_array,
-                                         ucp_dt_iov_t* iovec, int iovcnt)
+UCS_F_ALWAYS_INLINE ucp_dt_iov_t* get_ucp_iov(JNIEnv *env,
+                                              jlongArray addr_array, jlongArray size_array,
+                                              int &iovcnt)
 {
+    iovcnt = env->GetArrayLength(addr_array);
+    ucp_dt_iov_t* iovec = (ucp_dt_iov_t*)ucs_malloc(sizeof(*iovec) * iovcnt, "JUCX iov vector");
+
```
Comment (yosefe) [inline bindings/java/src/main/native/jucx_common_def.h:120]: remove space line

File: bindings/java/src/main/native/worker.cc [hunk 8]
```diff
@@ -164,10 +164,8 @@ Java_org_openucx_jucx_ucp_UcpWorker_recvTaggedIovNonBlockingNative(JNIEnv *env,
                                                                    jlong tag, jlong tag_mask,
                                                                    jobject callback)
 {
-    jsize iovcnt = env->GetArrayLength(addresses);
-    ucp_dt_iov_t* iovec = (ucp_dt_iov_t*)ucs_malloc(sizeof(ucp_dt_iov_t) * iovcnt, "JUCX iov vector");
-
-    fill_iov_vector(env, addresses, sizes, iovec, iovcnt);
+    int iovcnt;
+    ucp_dt_iov_t* iovec = get_ucp_iov(env, addresses, sizes, iovcnt);
```
Comment (yosefe) [inline bindings/java/src/main/native/worker.cc:168]: in all usages of get_ucp_iov:
if (iovec == NULL) 
 return

**Pull request merged**